dist/log.py

```diff
@@ -6,390 +6,652 @@
 # to be the metadata.
 def check_redact(optype):
-    for f  in optype.fields:
-        if f[0] == 'uint32_id':
-            redact_str = '\tif (!FLD_ISSET(args->flags, WT_TXN_PRINTLOG_UNREDACT) && '
-            redact_str += '%s != WT_METAFILE_ID)\n' % (f[1])
-            redact_str += '\t\treturn(__wt_fprintf(session, args->fs, " REDACTED"));\n'
+    for f in optype.fields:
+        if f.typename == 'uint32_id':
+            redact_str = '    if (!FLD_ISSET(args->flags, WT_TXN_PRINTLOG_UNREDACT) && '
+            redact_str += '%s != WT_METAFILE_ID)\n' % (f.fieldname)
+            redact_str += '        return(__wt_fprintf(session, args->fs, " REDACTED"));\n'
             return redact_str
     return ''
 
+
 # Create a printf line, with an optional setup function.
-# ishex indicates that the the field name in the output is modified
+# ishex indicates that the field name in the output is modified
 # (to add "-hex"), and that the setup and printf are conditional
 # in the generated code.
 def printf_line(f, optype, i, ishex):
     ifbegin = ''
     ifend = ''
-    nl_indent = '\n\t'
-    name = f[1]
+    nl_indent = '\n    '
+    name = f.fieldname
     postcomma = '' if i + 1 == len(optype.fields) else ',\\n'
     precomma = ''
     if ishex > 0:
         name += '-hex'
-        if not unconditional_hex(f):
+        if not f.always_hex:
             ifend = nl_indent + '}'
-            nl_indent += '\t'
+            nl_indent += '    '
             ifbegin = \
                 'if (FLD_ISSET(args->flags, WT_TXN_PRINTLOG_HEX)) {' + nl_indent
             if postcomma == '':
                 precomma = ',\\n'
     body = '%s%s(__wt_fprintf(session, args->fs,' % (
-        printf_setup(f, ishex, nl_indent),
+        f.printf_setup(ishex, nl_indent),
         'WT_ERR' if has_escape(optype.fields) else 'WT_RET') + \
         '%s    "%s        \\"%s\\": %s%s",%s));' % (
-        nl_indent, precomma, name, printf_fmt(f, ishex), postcomma,
-        printf_arg(f))
+        nl_indent, precomma, name, f.printf_fmt(ishex), postcomma,
+        f.printf_arg)
     return ifbegin + body + ifend
 
 #####################################################################
 # Create log_auto.c with handlers for each record / operation type.
 #####################################################################
-f='../src/log/log_auto.c'
-tfile = open(tmp_file, 'w')
 
-tfile.write('/* DO NOT EDIT: automatically built by dist/log.py. */\n')
+def run():
+    f='../src/log/log_auto.c'
+    tfile = open(tmp_file, 'w')
+
+    tfile.write('''/* DO NOT EDIT: automatically built by dist/log.py. */
 
-tfile.write('''
 #include "wt_internal.h"
 
+#define WT_SIZE_CHECK_PACK_PTR(p, end)     WT_RET_TEST(!(p) || !(end) || (p) >= (end), ENOMEM)
+#define WT_SIZE_CHECK_UNPACK_PTR(p, end)   WT_RET_TEST(!(p) || !(end) || (p) >= (end), EINVAL)
+#define WT_SIZE_CHECK_UNPACK_PTR0(p, end)  WT_RET_TEST(!(p) || !(end) || (p) >  (end), EINVAL)
```

Comment:
what is the need for the above two unpack check macros? Doesn't a single one satisfy the requirement?

Reply:
The ">=" versions check that there is more space in the buffer (i.e. at least one byte more left), the ">" one checks that the pointer has not gone beyond the end of the buffer (i.e. allows that 0 bytes are left).

Reply:
Thanks.